doc/developer/design/20230921_distributed_ts_oracle.md

```diff
@@ -0,0 +1,442 @@
+process fencing out other processes. We would also have to keep doing a
+`confirm_leadership()` at some point before returning query results.
+
+The distributed serving-layer processes would do timestamp operations via RPC to
+the Coordinator.
+
+### Add a separate (highly available) TimestampOracle service
+
+We would essentially pull the current in-memory/backed-by-stash implementation
+out into it's own service that clients talk to via (g)RPC. However, we can
+either accept that it is not highly available or we would have to do work that
+is similar to the current `confirm_leadership()` checks to ensure that operations
+return a linearizable timestamp. Operating a distributed, highly available
+service is hard, and by using CRDB (our main proposal) we delegate the hard work
+to that. Plus, we don't want to be operating yet more services as part of a
+customer deployment.
+
+## Open questions
+
+### What SQL type to use for the read/write timestamp?
+
+- `bigint`, so a rust `i64`: This goes until `+9223372036854775807`, which is
+  roughly `AD 292278994`.
+- `timestamp`, which goes until `AD 294276`. This would give us microsecond
+  resolution so we could squeeze in more than 1000 timestamps per second (which
+  is what we currently get), if we ever wanted to.
+
+TODO: I have yet to run benchmarks to see how performant the
+`read_ts`/`write_ts` queries become when arithmetic uses `timestamp` and
+`interval`.
+
+### A separate TimestampOracle for syncing Catalog state?
+
+In the future, multiple serving-layer processes listen to differential changes
+to Catalog state and we have to have a method for knowing if we are up to date.
+We can either use a `ReadHandle::fetch_recent_upper()` that guarantees
+linearizability, _or_ we could use a TimestampOracle.
+
+If we use a TimestampOracle, we could either use a completely separate oracle
+for Catalog state or we could re-use the TimestampOracle that we use for the
+real-time timeline but add another time dimension to it. For example, `read_ts`
+would become `(table_read_ts, catalog_read_ts)`.
+
+The benefit of the multi-dimensional timestamps approach is that we need fewer
+oracle operations and that the timestamps that are mostly used together are
+stored in one CRDB row. However, it does not provide good separation/doesn't
+seem a good abstraction. Additionally, we would have to make that call now or
+add the second dimension later, with a migration. The risk with adding this now
+is that we might find out later that we added the wrong thing.
+
+Storing Catalog timestamps in a separate oracle seems like the right
+abstraction but we have to make more oracle operations. The good thing is that
+timestamp operations can be pipelined/we can make requests to both oracles
+concurrently. It _does_ mean, however, that a single read now needs two oracle
+operations: one for getting the Catalog `read_ts` and one for getting the
+real-time `read_ts`.
```

Comment:
+1, I agree that this sounds like the right approach. I don't think `SELECT read_ts FROM timestamp_oracle WHERE timeline = $timeline;` Is going to be much faster than a query like `SELECT read_ts FROM timestamp_oracle WHERE timeline IN [$real_timeline, $oracle_timeline];`

Reply:
Currently, a `TimestampOracle` is scoped to a `timeline`, and that's how the coordinator uses it. We can change that in the future and say that `timeline` is passed in as a parameter, and that would also allow us to add, say, a `get_read_ts_multi(timelines: &[String])`. Then we could use one `SELECT` query to get multiple timestamps.

Is that what you had in mind? I hadn't thought about that before but think it's an excellent idea. I wouldn't change the `TimestampOracle` interface right now, though, because it requires more refactorings in the places where the oracle is used. The data model (using one `timestamp_oracle` table) would be ready for that, though, so it's easy to change the interface in the future. wdyt?

Reply:
> Is that what you had in mind?

Yes, that's what I was thinking. If we go with the catalog timeline approach, then I think 100% of the read and write queries will need both a user timeline timestamp and a catalog timeline timestamp. So it probably makes sense to combine the call into a single query. Maybe even a method like `get_timeline_and_catalog_read_ts(timeline: String)` if you wanted something less general. Either way works though.

> I wouldn't change the TimestampOracle interface right now, though, because it requires more refactorings in the places where the oracle is used. The data model (using one timestamp_oracle table) would be ready for that, though, so it's easy to change the interface in the future. wdyt?

I think that's a good approach. It makes sense to save it for later since it's only a performance optimization of something that doesn't even exist yet.

Reply:
Yeah, I like the "catalog is a timeline" approach a lot.

> However, it does not provide good separation/doesn't seem a good abstraction.

> I'm proposing that we store the Catalog timestamp in a separateTimestampOracle, if we decide to use oracles at all.

If that doesn't work out for some reason, IMO doing one query instead of two outweights the abstraction awkwardness. Two means more tail latency and higher crdb costs and the abstraction leakage would be minimal.